cairo-contracts/cometbft/src/ics23.cairo

```diff
@@ -0,0 +1,220 @@
+            HashOp::Sha512 => 2,
+            HashOp::Keccak256 => 3,
+            HashOp::Ripemd160 => 4,
+            HashOp::Bitcoin => 5,
+            HashOp::Sha512_256 => 6,
+            HashOp::Blake2b_512 => 7,
+            HashOp::Blake2b_256 => 8,
+            HashOp::Blake3 => 9,
+        }
+    }
+}
+
+impl U64IntoHashOp of Into<u64, HashOp> {
+    fn into(self: u64) -> HashOp {
+        match self {
+            0 => HashOp::NoOp,
+            1 => HashOp::Sha256,
+            2 => HashOp::Sha512,
+            3 => HashOp::Keccak256,
+            4 => HashOp::Ripemd160,
+            5 => HashOp::Bitcoin,
+            6 => HashOp::Sha512_256,
+            7 => HashOp::Blake2b_512,
+            8 => HashOp::Blake2b_256,
+            9 => HashOp::Blake3,
+            _ => panic!("invalid HashOp"),
+        }
+    }
+}
+
+
+#[derive(Default, Debug, Copy, Drop, PartialEq, Serde)]
+pub enum LengthOp {
+    #[default]
+    NoPrefix,
+    VarProto,
+    VarRlp,
+    Fixed32Big,
+    Fixed32Little,
+    Fixed64Big,
+    Fixed64Little,
+    Require32Bytes,
+    Require64Bytes,
+}
+
+impl LengthOpIntoU64 of Into<LengthOp, u64> {
+    fn into(self: LengthOp) -> u64 {
+        match self {
+            LengthOp::NoPrefix => 0,
+            LengthOp::VarProto => 1,
+            LengthOp::VarRlp => 2,
+            LengthOp::Fixed32Big => 3,
+            LengthOp::Fixed32Little => 4,
+            LengthOp::Fixed64Big => 5,
+            LengthOp::Fixed64Little => 6,
+            LengthOp::Require32Bytes => 7,
+            LengthOp::Require64Bytes => 8,
+        }
+    }
+}
+
+impl U64IntoLengthOp of Into<u64, LengthOp> {
+    fn into(self: u64) -> LengthOp {
+        match self {
+            0 => LengthOp::NoPrefix,
+            1 => LengthOp::VarProto,
+            2 => LengthOp::VarRlp,
+            3 => LengthOp::Fixed32Big,
+            4 => LengthOp::Fixed32Little,
+            5 => LengthOp::Fixed64Big,
+            6 => LengthOp::Fixed64Little,
+            7 => LengthOp::Require32Bytes,
+            8 => LengthOp::Require64Bytes,
+            _ => panic!("invalid length op"),
+        }
+    }
+}
+
+#[derive(Default, Debug, Clone, Drop, PartialEq, Serde)]
+pub struct InnerSpec {
+    pub child_order: Array<i32>,
+    pub child_size: i32,
+    pub min_prefix_length: i32,
+    pub max_prefix_length: i32,
+    pub empty_child: ByteArray,
+    pub hash: HashOp,
+}
+
+impl InnerSpecAsProtoMessage of ProtoMessage<InnerSpec> {
+    fn encode_raw(self: @InnerSpec, ref output: ByteArray) {
+        ProtoCodecImpl::encode_repeated_field(1, self.child_order, ref output);
+        ProtoCodecImpl::encode_field(2, self.child_size, ref output);
+        ProtoCodecImpl::encode_field(3, self.min_prefix_length, ref output);
+        ProtoCodecImpl::encode_field(4, self.max_prefix_length, ref output);
+        ProtoCodecImpl::encode_field(5, self.empty_child, ref output);
+        ProtoCodecImpl::encode_field(6, self.hash, ref output);
```

Comment:
Do you think it makes sense to have a `field_number` field within the `ProtoCodec` struct and expose `init()` and `finalize()` methods by the `ProtoCodecTrait` to eliminate the need for the `field_number` arguments in places like here altogether?

Reply:
The field numbers are associated with the struct fields - part of the struct definition, which should have been auto-generated. The field numbers don't need to be sequential - only strictly monotonic. 

Prost uses macro for this, where each generated field is tagged with its corresponding protobuf details, e.g. field number, types, repeated or not, optional or not.

Reply:
Ah I can apply your suggestion for input bytes and index - to reduce arguments in the encode/decode methods.

For decoding,
```cairo
decode_instance.branch(limit);
..
decode_instance.field(5, ref self.empty_child);
...
decode_instance.done();
```

For encoding,
```cairo
...
encode_instance.field(5, self.empty_child)
...
```